File: hphp/runtime/ext/sb_profile/ext_sb_profile.cpp
#include "hphp/runtime/ext/extension.h"
#include "hphp/runtime/vm/jit/prof-data-serialize.h"

namespace HPHP {

OptString HHVM_FUNCTION(sb_profile_ser,
                     const OptString& sb_root,
                     const OptString& prof_path) {
  auto const status = jit::serializeSBProfData(sb_root.data(), prof_path.data());
  return OptString{status.c_str()};
}

OptString HHVM_FUNCTION(sb_profile_deser,
                     const OptString& sb_root,
                     const OptString& prof_path,
                     int64_t warmup) {
  auto const flags = static_cast<jit::SBWarmupFlags>(warmup);
  auto const status =
    jit::deserializeSBProfData(sb_root.data(), prof_path.data(), flags);
  return OptString{status.c_str()};
}

static struct SbProfileExtension final : Extension {
  SbProfileExtension()
    : Extension("sb_profile", NO_EXTENSION_VERSION_YET, "hphp_hphpi") {}
  void moduleRegisterNative() override {
    HHVM_FE(sb_profile_ser);
    HHVM_FE(sb_profile_deser);
  }
} s_sb_profile_extension;
}

File: hphp/runtime/ext/sb_profile/ext_sb_profile.php
<?hh

/**
 * Bitmask flags controlling which warmup phases run during sandbox jumpstart
 * deserialization. Combine with bitwise OR to enable multiple phases.
 *
 * Values match the C++ SBWarmupFlags enum in prof-data-serialize.h.
 */
enum SBWarmupFlag: int {
  PRELOAD = 1;  // Parallel unit preloading
  JIT     = 2;  // Enqueue translations for async JIT compilation
  APC     = 4;  // APC key warmup via ExtensionRegistry data
  ALL     = 7;  // All phases enabled (PRELOAD | JIT | APC)
}

/**
 * Serializes sandbox profile. A profile can be serialized only once for the
 * lifetime of the HHVM process.
 *
 * @param string $sb_root - PHP root directory of the sandbox.
 * @param string $prof_path - Full path of the profile file.
 * @return string - Status of the serialization attempt.
 */
<<__Native>>
function sb_profile_ser(string $sb_root, string $prof_path): ?string;
/**
 * Deserializes sandbox profile. A profile can be deserialized only once for the
 * lifetime of the HHVM process.
 *
 * @param string $sb_root - PHP root directory of the sandbox.
 * @param string $prof_path - Full path of the profile file.
 * @param int $warmup - Bitmask of SBWarmupFlag values controlling which warmup
 *   phases to enable. Defaults to SBWarmupFlag::ALL (all phases).
 * @return string - Status of the deserialization attempt.
 */
<<__Native>>
function sb_profile_deser(
  string $sb_root,
  string $prof_path,
  /* SBWarmupFlag bitmask */ int $warmup = 7 /* SBWarmupFlag::ALL */,
): ?string;
